kotlinx-coroutines-core/common/src/flow/operators/Transform.kt

```diff
@@ -97,3 +99,44 @@ public fun <T> Flow<T>.onEach(action: suspend (T) -> Unit): Flow<T> = flow {
         emit(value)
     }
 }
+
+/**
+ * Reduces the given flow with [operation], emitting every intermediate result, including initial value.
+ * The first element is takes as initial value for operation accumulator.
+ * For example:
+ * ```
+ * flowOf(1, 2, 3, 4).scan { (v1, v2) -> v1 + v2 }.toList()
+ * ```
+ * will produce `[1, 3, 6, 10]`
+ */
+@FlowPreview
+public fun <T> Flow<T>.scan(operation: suspend (accumulator: T, value: T) -> T): Flow<T> = flow {
+    var accumulator: Any? = NullSurrogate
+    collect { value ->
+        accumulator = if (accumulator === NullSurrogate) {
+            value
+        } else {
+            operation(accumulator as T, value)
+        }
+        emit(accumulator as T)
+    }
+}
+
+/**
+ * Reduces the given flow with [operation], emitting every intermediate result, including initial value.
+ * An initial value is provided lazily by [initialSupplier] and is always immediately emitted.
+ * For example:
+ * ```
+ * flowOf(1, 2, 3).scan(::emptyList) { acc: List<Int>, value -> acc + value }.toList()
+ * ```
+ * will produce `[], [1], [1, 2], [1, 2, 3]]`.
+ */
+@FlowPreview
+public fun <T, R> Flow<T>.scan(initialSupplier: () -> R, @BuilderInference operation: suspend (accumulator: R, value: T) -> R): Flow<R> = flow {
```

Comment:
`suspend () -> R` ?  

Anyway, we need to consider its consistency with https://youtrack.jetbrains.com/issue/KT-7657
One comment worth noting here:

> To be consistent with existing fold and reduce functions we need to name variants of scan or accumulate with and without initial value differently.

Unlike a stock scan, this variant's use-case is highly questionable, so I propose to avoid it for now altogether.